yargs.js

```diff
@@ -716,139 +753,147 @@ function Yargs (processArgs, cwd, parentRequire) {
+          self.showHelp('log')
+          if (exitProcess) {
+            exitEarly(0)
+          }
+        } else if (key === versionOpt && argv[key]) {
+          if (exitProcess) setBlocking(true)
 
-        skipValidation = true
-        self.showHelp('log')
-        if (exitProcess) {
-          process.exit(0)
+          skipValidation = true
+          usage.showVersion()
+          if (exitProcess) {
+            exitEarly(0)
+          }
         }
-      } else if (key === versionOpt && argv[key]) {
-        if (exitProcess) setBlocking(true)
+      })
 
-        skipValidation = true
-        usage.showVersion()
-        if (exitProcess) {
-          process.exit(0)
-        }
+      // Check if any of the options to skip validation were provided
+      if (!skipValidation && options.skipValidation.length > 0) {
+        skipValidation = Object.keys(argv).some(function (key) {
+          return options.skipValidation.indexOf(key) >= 0 && argv[key] === true
+        })
       }
-    })
 
-    // Check if any of the options to skip validation were provided
-    if (!skipValidation && options.skipValidation.length > 0) {
-      skipValidation = Object.keys(argv).some(function (key) {
-        return options.skipValidation.indexOf(key) >= 0 && argv[key] === true
-      })
-    }
-
-    // If the help or version options where used and exitProcess is false,
-    // or if explicitly skipped, we won't run validations
-    if (!skipValidation) {
-      if (parsed.error) throw parsed.error
-
-      // if we're executed via bash completion, don't
-      // bother with validation.
-      if (!argv[completion.completionKey]) {
-        validation.nonOptionCount(argv)
-        validation.missingArgumentValue(argv)
-        validation.requiredArguments(argv)
-        if (strict) validation.unknownArguments(argv, aliases)
-        validation.customChecks(argv, aliases)
-        validation.limitedChoices(argv)
-        validation.implications(argv)
+      // If the help or version options where used and exitProcess is false,
+      // or if explicitly skipped, we won't run validations
+      if (!skipValidation) {
+        if (parsed.error) throw parsed.error
+
+        // if we're executed via bash completion, don't
+        // bother with validation.
+        if (!argv[completion.completionKey]) {
+          validation.nonOptionCount(argv)
+          validation.missingArgumentValue(argv)
+          validation.requiredArguments(argv)
+          if (strict) validation.unknownArguments(argv, aliases)
+          validation.customChecks(argv, aliases)
+          validation.limitedChoices(argv)
+          validation.implications(argv)
+        }
       }
+    } catch (e) {
+      if (e === ExitEarly) exitFn(ExitEarly.code)
+      else throw e
     }
 
-    setPlaceholderKeys(argv)
-
-    return argv
+    // if a custom exit handler has been provided,
+    // for API consistency always trigger it.
+    if (exitFn !== process.exit) exitFn(0)
```

Comment:
This is called before validation of positional args for a command, so the current `exiting` check isn't quite sufficient (since, if positional arg validation fails, the custom exit handler would be given a code of 0).